README.md
# b.b.m.2 - metronome

b.b.m.2 is a metronome web application designed to help musicians keep a steady tempo while practicing or performing. It produces an audible sound at regular intervals.<br><br>
The tool uses HTML, CSS, and JavaScript to render a user interface that includes a large display of the current tempo, buttons to adjust the tempo, and buttons to start/stop/tap the metronome. The user can also use the "Tap Tempo" button to manually set the desired beat when tapping the button 4 or more consecutive times.
<br><br>
The CSS code provides a responsive design that adapts to various screen sizes and sets the style for all of the user interface elements. The JavaScript code defines variables and functions to manage the behavior of the metronome tool. The metronome sound is generated using the Web Audio API, which creates an oscillator object that generates a tone at a specific frequency for a specified duration.

## Features

- **Set BPM directly or increment/decrement it using buttons:** Users can enter a BPM value directly into an input field or use "+" and "-" buttons to increment/decrement the BPM value.
- **Calculate BPM based on user taps:** Users can tap a "Tap Tempo" button in time with their desired tempo and the metronome will calculate and set the corresponding BPM value.
- **Audible sound for each beat:** The metronome uses an oscillator from the Web Audio API to produce an audible sound for each beat.
- **Visual aids to represent the BPM:** A circle visual indicator changes opacity in time with the metronome beats to provide a visual representation of the tempo.

## Technical Details

The metronome is implemented using HTML, CSS, and JavaScript. JavaScript is used in several ways, such as:

- Adding event listeners to buttons and other elements to respond to user interactions
- Using functions to create timers for clearing old tap times and playing beats at regular intervals
- Using the Web Audio API to create an oscillator and control its frequency and volume
- Creating and syncing visual elements to the tempo beep

### The code is designed to be simple yet fine-tuned to deliver crucial functionality, ergonomy, and visual accessibility. It's a useful tool for musicians as well as a valuable resource for developers looking to study how such an application can be implemented.

#### This is a personal project and exercise. While the primary aim is to serve personal purposes, the process, code, and deployment are welcomed to be used as a tool in deployed form and as study material in code form. I hope you find it useful!

# b.b.m.2 can be accessed <a href="https://retzilience.github.io/beep-beep-machine/" target="_blank">here</a></h2>

## the javascript is extensively commented <a href="https://github.com/retzilience/beep-beep-machine/blob/main/comment/script.js" target="_blank">here</a></h2>

### you can also find the commented style.css and index.html <a href="https://github.com/retzilience/beep-beep-machine/tree/main/comment" target="_blank">here</a>

#### additionally, if you haven't had enough of your fill of metronome code, you can find the project table of contents <a href="https://github.com/retzilience/beep-beep-machine/blob/main/toc.md">here</a>

<br>

<h1>これはメトロノームです</h1>

<p>メトロノームは、演奏や練習中に音楽家が一定のテンポを保つのに役立つ道具です。 定期的な間隔で聴覚的な音を出し、ビート/分（BPM）の数はユーザーによって調整可能です。</p>

<h2>いくつかの機能が含まれています</h2>
<ul>
  <li>BPMを直接設定するか、ボタンを使用して増減する機能</li>
  <li>ユーザーのタップに基づいてBPMを計算する機能</li>
  <li>各ビートごとに聴覚的な音を生成するためのオシレーターの使用</li>
</ul>

<h2>JavaScriptの使用</h2>
<p>JavaScriptは、次の形式で複数使用されます。</p>
<ul>
  <li>ボタンやその他の要素にイベントリスナーを追加して、ユーザー操作に応答します。</li>
  <li>setTimeoutおよびsetIntervalを使用して、古いタップ時間をクリアし、定期的な間隔でビートを再生するためのタイマーを作成します。</li>
  <li>Web Audio API を使用してオシレーターを作成し、その周波数と音量を制御します。</li>
</ul>

<p>これは個人的なプロジェクトとエクササイズです。 主な目的は個人的な目的に役立つことですが、プロセス、コード、およびデプロイメントはデプロイされた形式でツールとして使用されることも歓迎されます。</p>



<h1>Este é um metrônomo</h1>

<p>O metrônomo é uma ferramenta que ajuda os músicos a manter um ritmo constante enquanto praticam ou se apresentam. Ele produz um som audível em intervalos regulares, com o número de batidas por minuto (BPM) sendo ajustável pelo usuário.</p>

<h2>Algumas das características que inclui</h2>
<ul>
  <li>A capacidade de definir o BPM diretamente ou incrementar/decrementar usando botões</li>
  <li>A capacidade de calcular o BPM com base nas batidas do usuário</li>
  <li>O uso de um oscilador para produzir um som audível para cada batida</li>
</ul>

<h2>Uso do JavaScript</h2>
<p>O JavaScript é usado em várias formas como:</p>
<ul>
  <li>Adicionar ouvintes de eventos a botões e outros elementos para responder às interações do usuário</li>
  <li>Usando setTimeout e setInterval para criar temporizadores para limpar tempos antigos de toque e tocar batidas em intervalos regulares</li>
  <li>Usando a API Web Audio para criar um oscilador e controlar sua frequência e volume</li>
</ul>

<p>Este é um projeto pessoal e exercício. Embora o objetivo principal seja atender a fins pessoais, o processo, código e implantação são bem-vindos para serem usados como uma ferramenta na forma implantada e como material de estudo na forma de código.</p>

<h1>Este es un metrónomo</h1>

<p>El metrónomo es una herramienta que ayuda a los músicos a mantener un tempo constante mientras practican o actúan. Produce un sonido audible en intervalos regulares, con el número de beats por minuto (BPM) ajustable por el usuario.</p>

<h2>Algunas de las características que incluye</h2>
<ul>
  <li>La capacidad de establecer el BPM directamente o incrementarlo/decrementarlo usando botones.</li>
  <li>La capacidad de calcular el BPM en función del toque del usuario.</li>
  <li>El uso de un oscilador para producir un sonido audible en cada beat.</li>
</ul>

<h2>Uso del JavaScript</h2>
<p>JavaScript se usa en varias formas como:</p>
<ul>
  <li>Agregar escuchas de eventos a botones y otros elementos para responder a las interacciones del usuario.</li>
  <li>Usando setTimeout y setInterval para crear temporizadores para borrar tiempos antiguos y reproducir beats en intervalos regulares.</li>
  <li>Usando la API Web Audio para crear un oscilador y controlar su frecuencia y volumen.</li>
</ul>

<p>Este es un proyecto personal y ejercicio. Si bien el objetivo principal es servir a propósitos personales, se da la bienvenida al proceso, código y despliegue como herramienta en forma desplegada y como material de estudio en forma de código.</p>



<h1>이것은 메트로놈입니다</h1>

<p>메트로놈은 연주나 연습 중 음악가들이 일정한 템포를 유지하는 데 도움이 되는 도구입니다. 정기적인 간격으로 청각적인 소리를 내며 사용자가 조절할 수 있는 분당 비트 수 (BPM) 입니다.</p>

<h2>포함하는 몇 가지 기능</h2>
<ul>
  <li>BPM을 직접 설정하거나 버튼을 사용하여 증가/감소시키는 기능</li>
  <li>사용자 탭에 따라 BPM을 계산하는 기능</li>
  <li>각 비트마다 청각적인 소리를 생성하기 위해 오실레이터 사용</li>
</ul>

<h2>JavaScript 사용</h2>
<p>JavaScript는 다음과 같은 여러 형태로 사용됩니다.</p>
<ul>
  <li>버튼 및 기타 요소에 이벤트 리스너 추가하여 사용자 상호 작용에 응답</li>
  <li>setTimeout 및 setInterval을 사용하여 이전 탭 시간을 지우고 정기적인 간격으로 비트 재생을위한 타이머 생성</li>
  <li>Web Audio API를 사용하여 오실레이터 생성 및 주파수 및 볼륨 제어</li>
</ul>

<p>개인 프로젝트와 운동입니다. 주요 목표는 개인적인 목적에 부합하는 것이지만 프로세스, 코드 및 배포는 배포된 형태의 도구로서와 코드 형태의 학습 자료로서의 사용도 환영합니다.</p>


changelog
# Changelog

## [2.02] - 2023-03-12

### Added
- Created stand-alone versions of the CSS and JavaScript files.

### Changed
- Updated the project structure from a single HTML page to an HTML file that calls the script and CSS separately.
- Updated the commented project files to reflect the new project structure.
- Updated the README file to include information on the new project structure and usage.

### Fixed
- N/A

## [2.01] - 2023-03-10
### Added
- Added a media query to apply the `user-select: none` property only on mobile devices. This allows interaction with form elements on desktop while still preventing text selection on mobile devices.

### Changed
- Changed CSS code to remove the up and down arrows from the number input box. This allows users to change the BPM value by typing only.

### Fixed
- Fixed an issue where text selection was disabled for all elements on the page, preventing interaction with form elements like input fields on desktop.

These changes improved the usability of the metronome by allowing users to interact with form elements on desktop and providing a cleaner interface for changing BPM values.


beta/betascript.js
document.addEventListener("gesturestart", function (event) {
  event.preventDefault();
});
document.addEventListener(
  "touchstart",
  function (event) {
    if (event.touches.length > 1) {
      event.preventDefault();
    }
  },
  { passive: false }
);
document.addEventListener("gesturestart", function (e) {
  e.preventDefault();
});
var audioContext = new AudioContext();
var gainNode = audioContext.createGain();
gainNode.connect(audioContext.destination);
var duration = 0.025;
var beat = audioContext.createOscillator();
beat.frequency.value = 1000;
beat.connect(gainNode);
beat.start();
beat.stop(audioContext.currentTime + duration);
var isPlaying = false;
var bpmInput = document.getElementById("bpm");
bpmInput.addEventListener("input", function () {
  var lastBpm = 0;
  bpmInput.addEventListener("change", function () {
    if (isPlaying && parseInt(this.value) !== lastBpm) {
      setTimeout(function () {
        var bpm = parseInt(bpmInput.value);
        var interval = 60000 / bpm;
        clearInterval(timer);
        timer = setInterval(function () {
          beat = audioContext.createOscillator();
          beat.frequency.value = 1000;
          beat.connect(gainNode);
          beat.start(audioContext.currentTime);
          beat.stop(audioContext.currentTime + duration);
          circle.style.opacity = "1.0";
          setTimeout(function () {
            circle.style.opacity = "0.0";
          }, interval / 2);
        }, interval);
      }, 500);
      lastBpm = parseInt(this.value);
    }
  });
  var bpm = parseInt(this.value);
  if (bpm >= 1 && bpm <= 300) {
    updateBpmValue(bpm);
  }
});
var startButton = document.getElementById("start");
var stopButton = document.getElementById("stop");
var circle = document.querySelector(".circle");
startButton.addEventListener("click", function () {
  if (!isPlaying) {
    var bpm = parseInt(bpmInput.value);
    var interval = 60000 / bpm;
    timer = setInterval(function () {
      beat = audioContext.createOscillator();
      beat.frequency.value = 1000;
      beat.connect(gainNode);
      beat.start(audioContext.currentTime);
      beat.stop(audioContext.currentTime + duration);
      circle.style.opacity = "1.0";
      setTimeout(function () {
        circle.style.opacity = "0.0";
      }, interval / 2);
    }, interval);
    isPlaying = true;
  }
});
stopButton.addEventListener("click", function () {
  if (isPlaying) {
    clearInterval(timer);
    isPlaying = false;
    circle.style.opacity = "0.0";
  }
});

function updateMetronomeInterval() {
  if (isPlaying) {
    clearInterval(timer);
    var bpm = parseInt(bpmInput.value);
    var interval = 60000 / bpm;
    timer = setInterval(function () {
      beat = audioContext.createOscillator();
      beat.frequency.value = 1000;
      beat.connect(gainNode);
      beat.start(audioContext.currentTime);
      beat.stop(audioContext.currentTime + duration);
      circle.style.opacity = "1.0";
      setTimeout(function () {
        circle.style.opacity = "0.0";
      }, interval / 2);
    }, interval);
  }
}
var tapButton = document.getElementById("tap");
var tapTimes = [];
var tapTimer;
tapButton.addEventListener("click", function () {
  tapTimes.push(Date.now());
  clearTimeout(tapTimer);
  if (tapTimes.length > 3) {
    tapTimes.shift();
    var total = 0;
    for (var i = 1; i < tapTimes.length; i++) {
      total += tapTimes[i] - tapTimes[i - 1];
    }
    var average = total / (tapTimes.length - 1);
    var bpm = Math.round(60000 / average);
    bpmInput.value = bpm;
    updateMetronomeInterval();
  }
  tapTimer = setTimeout(function () {
    tapTimes = [];
  }, 2000);
});
var bpmDecreaseButton = document.getElementById("bpm-decrease");
var bpmIncreaseButton = document.getElementById("bpm-increase");
bpmDecreaseButton.addEventListener("click", function () {
  var bpm = parseInt(bpmInput.value);
  if (bpm > 1) {
    bpmInput.value = bpm - 1;
    updateMetronomeInterval();
  }
});
bpmIncreaseButton.addEventListener("click", function () {
  var bpm = parseInt(bpmInput.value);
  if (bpm < 300) {
    bpmInput.value = bpm + 1;
    updateMetronomeInterval();
  }
});
function updateBpmValue(value) {
  bpmInput.value = value;
}
bpmInput.addEventListener("click", function () {
  this.focus();
});


comment/script.js
// Prevent the default behavior of certain touch and gesture events to maintain a good and reliable user interface that doesn’t interfere with the usage of the metronome.
document.addEventListener("gesturestart", function (event) {
  event.preventDefault();
});
document.addEventListener(
  "touchstart",
  function (event) {
    if (event.touches.length > 1) {
      event.preventDefault();
    }
  },
  { passive: false }
);
document.addEventListener("gesturestart", function (e) {
  e.preventDefault();
});

// An AudioContext object is created and used to create an oscillator node (beat) and a gain node (gainNode). The gain node is connected to the audio context’s destination (the speakers), and the oscillator node is connected to the gain node. The oscillator produces a sound at a frequency of 1000 Hz for a duration of 0.025 seconds.

// Create an audio context and gain node for playing the metronome sound
var audioContext = new AudioContext();
var gainNode = audioContext.createGain();
gainNode.connect(audioContext.destination);

// Set the duration of the metronome sound
var duration = 0.025;

// Create an oscillator for playing the metronome sound
var beat = audioContext.createOscillator();
beat.frequency.value = 1000;
beat.connect(gainNode);
beat.start();
beat.stop(audioContext.currentTime + duration);

// Keep track of whether the metronome is currently playing
var isPlaying = false;

// Get references to DOM elements for user input and interaction
var bpmInput = document.getElementById("bpm");
var startButton = document.getElementById("start");
var stopButton = document.getElementById("stop");
var circle = document.querySelector(".circle");
// Sets up event listeners for various user interface elements such as buttons for starting and stopping the metronome, adjusting the beats per minute (BPM), and tapping to set the BPM.
// When the start button is clicked, if the metronome is not already playing, an interval timer is set up based on the current BPM value. At each interval, an oscillator node is created and started to produce a sound at 1000 Hz for 0.025 seconds. A visual indicator (circle) also briefly changes opacity.
// When the stop button is clicked, if the metronome is playing, it stops by clearing the interval timer.
// Update BPM value when user changes input field value
// If the metronome is playing and the new BPM value is different from the previous one, update the interval timer to reflect the new BPM value
// A timeout is added to prevent immediate changes in BPM from interfering with the current interval timer
bpmInput.addEventListener("input", function () {
  var lastBpm = 0;
  bpmInput.addEventListener("change", function () {
    if (isPlaying && parseInt(this.value) !== lastBpm) {
      setTimeout(function () {
        var bpm = parseInt(bpmInput.value);
        var interval = 60000 / bpm;
        clearInterval(timer);
        timer = setInterval(function () {
          beat = audioContext.createOscillator();
          beat.frequency.value = 1000;
          beat.connect(gainNode);
          beat.start(audioContext.currentTime);
          beat.stop(audioContext.currentTime + duration);
          circle.style.opacity = "1.0";
          setTimeout(function () {
            circle.style.opacity = "0.0";
          }, interval / 2);
        }, interval);
      }, 500);
      lastBpm = parseInt(this.value);
    }
  });
  var bpm = parseInt(this.value);
  if (bpm >= 1 && bpm <= 300) {
    updateBpmValue(bpm);
  }
});
// Get the start, stop buttons and circle element from the DOM

var startButton = document.getElementById("start");
var stopButton = document.getElementById("stop");
var circle = document.querySelector(".circle");


// Start playing the metronome when start button is clicked.
startButton.addEventListener("click", function () {
  if (!isPlaying) {
    var bpm = parseInt(bpmInput.value);
    var interval = 60000 / bpm;
    timer = setInterval(function () {
      beat = audioContext.createOscillator();
      beat.frequency.value = 1000;
      beat.connect(gainNode);
      beat.start(audioContext.currentTime);
      beat.stop(audioContext.currentTime + duration);
      // This lights up the circle and sets a timeout with the interval to half the time of the beat, meaning the circle will become transparent again halfway through the current beat.
      circle.style.opacity = "1.0";
      setTimeout(function () {
        circle.style.opacity = "0.0";
      }, interval / 2);
    }, interval);
    // Indicates that the metronome started playing
    isPlaying = true;
  }
});

// Stop playing the metronome when stop button is clicked
stopButton.addEventListener("click", function () {
  if (isPlaying) {
    clearInterval(timer);
    isPlaying = false;
    circle.style.opacity = "0.0";
  }
});
//
// This function updates the metronome's interval based on the user's input
// It clears the current timer and sets a new one with the updated interval
function updateMetronomeInterval() {
  if (isPlaying) {
    clearInterval(timer);
    var bpm = parseInt(bpmInput.value);
    var interval = 60000 / bpm;
    timer = setInterval(function () {
      beat = audioContext.createOscillator();
      beat.frequency.value = 1000;
      beat.connect(gainNode);
      beat.start(audioContext.currentTime);
      beat.stop(audioContext.currentTime + duration);
      circle.style.opacity = "1.0";
      setTimeout(function () {
        circle.style.opacity = "0.0";
      }, interval / 2);
    }, interval);
  }
}
// The tap button allows users to tap in time with their desired BPM. Each time it’s clicked, it records the current time in an array (tapTimes). If more than two taps have been recorded within two seconds (determined by a timeout), it calculates an average interval between taps and sets the BPM value accordingly.
// The tapButton variable is assigned to the element with the ID “tap”. An empty array tapTimes and a variable tapTimer are also declared.
// An event listener is added to the tap button for click events. When clicked, the current time in milliseconds since January 1, 1970 (obtained using Date.now()) is pushed to the tapTimes array. Any existing timeout set by tapTimer is cleared using clearTimeout().
// If there are more than two times in the tapTimes array, it calculates an average interval between taps. This is done by looping through the array starting from index 1 and adding up the differences between each time and its previous time. The total difference is then divided by one less than the number of times to get an average interval in milliseconds.
// The BPM value is then calculated by dividing 60000 (the number of milliseconds in a minute) by this average interval and rounding to the nearest integer. The BPM input field’s value is updated with this calculated BPM.
// A timeout is set using setTimeout() to clear out the tapTimes array after two seconds have passed since the last tap. This allows users to start over with a new series of taps if they pause for more than two seconds between taps.
var tapButton = document.getElementById("tap");
var tapTimes = [];
var tapTimer;
tapButton.addEventListener("click", function () {
  tapTimes.push(Date.now());
  clearTimeout(tapTimer);
  if (tapTimes.length > 3) {
    tapTimes.shift();
    var total = 0;
    for (var i = 1; i < tapTimes.length; i++) {
      total += tapTimes[i] - tapTimes[i - 1];
    }
    var average = total / (tapTimes.length - 1);
    var bpm = Math.round(60000 / average);
    bpmInput.value = bpm;
    updateMetronomeInterval();
  }
  tapTimer = setTimeout(function () {
    tapTimes = [];
  }, 2000);
});

// Increase and decrease BPM value using buttons
var bpmDecreaseButton = document.getElementById("bpm-decrease");
var bpmIncreaseButton = document.getElementById("bpm-increase");
bpmDecreaseButton.addEventListener("click", function () {
  var bpm = parseInt(bpmInput.value);
  if (bpm > 1) {
    bpmInput.value = bpm - 1;
    updateMetronomeInterval();
  }
});
bpmIncreaseButton.addEventListener("click", function () {
  var bpm = parseInt(bpmInput.value);
  if (bpm < 300) {
    bpmInput.value = bpm + 1;
    updateMetronomeInterval();
  }
});

// Update BPM input field value
function updateBpmValue(value) {
  bpmInput.value = value;
}

// Focus BPM input field when clicked (useful on mobile)
bpmInput.addEventListener("click", function () {
  this.focus();
});
